eclair-front/src/main/scala/fr/acinq/eclair/router/FrontRouter.scala

```diff
@@ -0,0 +1,320 @@
+                  origin.peerConnection ! TransportHandler.ReadAck(ann)
+                  Metrics.gossipStashedRebroadcast(ann).increment()
+                  d.copy(rebroadcast = d.rebroadcast.copy(nodes = d.rebroadcast.nodes + (n -> (d.rebroadcast.nodes(n) + origin))))
+                case c: ChannelAnnouncement if d.rebroadcast.channels.contains(c) =>
+                  origin.peerConnection ! TransportHandler.ReadAck(ann)
+                  Metrics.gossipStashedRebroadcast(ann).increment()
+                  d.copy(rebroadcast = d.rebroadcast.copy(channels = d.rebroadcast.channels + (c -> (d.rebroadcast.channels(c) + origin))))
+                case u: ChannelUpdate if d.rebroadcast.updates.contains(u) =>
+                  origin.peerConnection ! TransportHandler.ReadAck(ann)
+                  Metrics.gossipStashedRebroadcast(ann).increment()
+                  d.copy(rebroadcast = d.rebroadcast.copy(updates = d.rebroadcast.updates + (u -> (d.rebroadcast.updates(u) + origin))))
+                case _ =>
+                  Metrics.gossipForwarded(ann).increment()
+                  log.info("sending announcement class={} to master router", ann.getClass.getSimpleName)
+                  remoteRouter ! PeerRoutingMessage(self, remoteNodeId, ann) // nb: we set ourselves as the origin
+                  d.copy(processing = d.processing + (ann -> Set(origin)))
+              }
+          }
+      }
+      stay using d1
+
+    case Event(accepted: GossipDecision.Accepted, d) =>
+      log.debug("message has been accepted by router: {}", accepted)
+      Metrics.gossipAccepted(accepted.ann).increment()
+      d.processing.get(accepted.ann) match {
+        case Some(origins) => origins.foreach { origin =>
+          log.debug("acking msg={} for origin={}", accepted.ann, origin)
+          origin.peerConnection ! TransportHandler.ReadAck(accepted.ann)
+          origin.peerConnection ! accepted
+        }
+        case None => ()
+      }
+      // NB: we will also shortly receive a NetworkEvent from the router for this announcement, where we put it in the
+      // rebroadcast map. We keep the origin peers in the accepted map, so we don't send back the same announcement to
+      // the peers that sent us in the first place.
+      // Why do we not just leave the announcement in the processing map? Because we would have a race between other
+      // peers that send us that same announcement, and the NetworkEvent. If the other peers win the race, we will defer
+      // acknowledging their message (because the announcement is still in the processing map) and we will
+      // wait forever for the very gossip decision that we are processing now, resulting in a stuck connection
+      val origins1 = d.processing.getOrElse(accepted.ann, Set.empty[RemoteGossip])
+      stay using d.copy(processing = d.processing - accepted.ann, accepted = d.accepted + (accepted.ann -> origins1))
```

Comment:
The `accepted` map will be cleaned up when receiving the `NetworkEvent` for that announcement. If that event never arrives, because of a bug, we will never free the memory.

We could choose to arbitrarily clean up this map when processing `TickBroadcast` events, but then we have a new race condition, where we regularly forget the origin information and send back announcements to peers that already know it.

Reply:
The current solution feels ok. We can always monitor memory consumption, that should indicate potential bugs.
And front machines should be expendable, so it's not a big issue if we have to restart one at some point because of a memory leak (and will help figure out the bug and fix it afterwards).